Enable Velox Parquet Reader with IAA

ruclz opened the pull request:
Issue link [4379] (https://github.com/facebookincubator/velox/issues/4379)
The Intel® In-Memory Analytics Accelerator (Intel® IAA) is a hardware accelerator that provides very high
throughput compression and decompression combined with primitive analytic functions. It is available in the newest generation of Intel® Xeon® Scalable processors ("Sapphire Rapids"). We observed a decompression hotspot like snappy::Rawuncompress function when Velox runs its TpchBenchmark with parquet format data . The IAA has a good chance to accelerate Velox, especially table scan operation intensive workload. Therefore, we enabled the IAA to Velox native parquet reader.

The IAA relies on [QPL](https://intel.github.io/qpl/index.html) library, the QPL only supports Linux systems and x86 processor. If the CPU has no IAA devices, the QPL can be run with software path, if CPU has, it can be run with hardware path.

We run the TpchBenchmark and SsbBenchmark, the SsbBenchmark is implemented by us, not including in this PR. The result shows.


- In terms of performance, IAA decompression has 9%~15% improvement compared with ZSTD, and is very close to snappy, 21%(SsbBenchmark) and 79%(TpchBenchmark) improvement compared with gzip.

- In terms if compression ratio, IAA has 5%~12% improvement compared with zstd, has 16%~21% improvement compared with snappy, and is very close to gzip.

In this PR, because IAA (de)compression needs initialize job buffer for submitting job to IAA, these job buffer need to free when Velox exits, but now I don`t know where to free. If you know, please tell me, thanks.

Yuhta commented on velox/dwio/parquet/thrift/ParquetThriftTypes.h:
```diff
@@ -318,7 +318,10 @@ struct CompressionCodec {
     ZSTD = 6,
-    LZ4_RAW = 7
+    LZ4_RAW = 7,
+#if defined(__linux__) && (defined(__x86_64__) || defined(__i386__))
```
We can keep the enum value unconditionally

Yuhta commented on CMakeLists.txt:
```diff
@@ -36,7 +36,7 @@ option(
   VELOX_BUILD_MINIMAL
   "Build a minimal set of components only. This will override other build options."
-  OFF)
+  ON)
```
Change the default value back

Yuhta commented on velox/dwio/parquet/reader/PageReader.cpp:
```diff
@@ -26,6 +26,10 @@
 #include <zlib.h>
 #include <zstd.h>
 
+#if defined(__linux__) && (defined(__x86_64__) || defined(__i386__))
```
Replace with `VELOX_ENABLE_QPL` once it's added

Yuhta commented on third_party/cmake_modules/FindThrift.cmake:
Merge conflict, just revert

ruclz commented on CMakeLists.txt:
```diff
@@ -366,6 +366,13 @@ if(NOT ${VELOX_BUILD_MINIMAL})
   find_library(SNAPPY snappy)
+  if((${CMAKE_SYSTEM_NAME} MATCHES "Linux") AND (${CMAKE_SYSTEM_PROCESSOR}
+                                                 MATCHES "x86_64"))
+    execute_process(COMMAND apt -y install uuid-dev nasm)
```
Of course, I can move it to setup shell, if this, the PR will not pass some ci check, such as linux build, linux-build-options

ruclz commented on CMake/FindICU.cmake:
Yes, but if this, the PR can't pass format check CI and don`t know why.

ruclz commented:
@Yuhta , thanks for your comments, I have synced the version with main branch of upstream. Also, I have added the 'VELOX_ENABLE_QPL' in the CMakelists.txt file, and modified related code in the cpp file. The nasm and uuid libraries dependency will be installed in setup scripts, including ubuntu, centos-8, circleci and adapter script.

Yuhta commented on velox/dwio/parquet/reader/PageReader.cpp:
```diff
@@ -190,6 +194,25 @@ const char* FOLLY_NONNULL PageReader::uncompressData(
+      dwio::common::ensureCapacity<char>(
+          uncompressedData_, uncompressedSize, &pool_);
+
+      Qplcodec* qpl_dec =
```
Just allocate this on stack.  We need to release the resources in all cases.

Yuhta commented on velox/dwio/parquet/thrift/ParquetThriftTypes.h:
```diff
@@ -318,7 +318,8 @@ struct CompressionCodec {
     ZSTD = 6,
-    LZ4_RAW = 7
+    LZ4_RAW = 7,
+    QPL = 8
```
nit: Leave a comma at end so next time we don't need to add it

ruclz commented on velox/dwio/parquet/reader/CompressionQpl.h:
```diff
@@ -0,0 +1,51 @@
+        job_(NULL),
+        idx_(0){};
+  ~Qplcodec() {
+    // Freejob();
```
@Yuhta Yeah, there is a memory leak problem. When velox calls QPL:decompress function first time, it need to malloc some job pool. So, the later function call will not malloc job pool, but when velox exits, we need to free these job pool. But at present, I don`t know where to free these job pool. It will be helpful if you can give some suggestions. I think, Velox also need to free some global memory space, the job pool free can be put there.

Yuhta replied:
How large is the malloc and initialization cost?  For other decompression we just release the memory on each invocation and usually this is not a bottleneck for performance.  Can you do a simple benchmark and see if it is feasible to do the same?

ruclz replied:
Thanks, the job pool size is about 120KB for hardware path or 700KB for software path. The initialization cost is high if we release the memory on each invocation. This finding has been proven on other workloads, such as oceanbase. If we don`t release these memory, [**job pool count * job pool size**] (KB) memory may lead to memory leak.

ruclz replied:
Thanks for your comment, it`s a good idea, I will have a try.

Yuhta commented on velox/dwio/parquet/reader/CompressionQpl.cpp:
```diff
@@ -0,0 +1,207 @@
+    qpl_status status = qpl_execute_job(job_);
+    if (status != QPL_STS_OK) {
+      std::atomic_store(&job_status[job_id], false);
+      VELOX_FAIL("QPL::Error while decompression occurred.");
```
We need to choose whether to return the success state or using exception, not both.  Can you
1. Choose whether we return value (remove all throws in this file and throw up in the stack), or just throw here (return `void` instead)
2. Decide if empty buffer is a valid case?  If not, we need to throw an exception for that (here or up in the stack)

ruclz replied:
Yeah, I have changed bool-return function to void function. And we throw an exception if buffer is empty.

Yuhta commented on velox/dwio/parquet/reader/CompressionQpl.cpp:
```diff
@@ -0,0 +1,151 @@
+
+thread_local bool initialize_job = false;
+thread_local uint8_t job_buffer[JOB_BUFFER_SIZE];
+thread_local qpl_job* job;
```
You can put all these inside `Qplcodec` and have `thread_local Qplcodec` here, then access it using a static method `static Qplcodec& Qplcodec::getInstance()`.  In `~Qplcodec()` you can release the resource.

ruclz commented:
@Yuhta Thanks， the job_buffer variable has been put in object, please help to review.

assignUser commented on velox/dwio/parquet/reader/CMakeLists.txt:
```diff
@@ -12,8 +12,17 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+if(${VELOX_ENABLE_QPL})
+  set(CompressionQpl "CompressionQpl.cpp")
+  set(CompressionQplDependency "qpl")
+else()
+  set(CompressionQpl "")
+  set(CompressionQplDependency "")
+endif()
+
```
Please avoid using variables for target related things. You can use [target_sources](https://cmake.org/cmake/help/latest/command/target_sources.html) to conditionally add sources to an existing target and just call `target_link_libraries` again to add any dependencies.

ruclz replied:
Thanks, I have completed it. Please help to review

ruclz commented on CMake/resolve_dependency_modules/Findqpl.cmake:
Thank you, can I delete qplfind.cmake file and force qpl from source (modify default qpl_SOURCE BUNDLED), it also can avoid some system build problems such as version.

ruclz replied:
Yes, now we force qpl library from source avoid to get system qpl library. In velox, only qpl version 1.1.0 can work. Is this ok for you? Thanks for your comment. @assignUser

Yuhta commented on velox/dwio/parquet/reader/CompressionQpl.h:
```diff
@@ -0,0 +1,49 @@
+      uint8_t* output);
+
+ private:
+  bool initialize_job;
```
`jobInitialized_`, also we are using camel cases for variable names, function names need to start with lower case

Yuhta reviewed (commented):
Can we add some test?  You can put the data in `tests/examples`

Yuhta commented on CMakeLists.txt:
```diff
@@ -178,6 +180,10 @@ endif()
 
 if(VELOX_ENABLE_PARQUET)
   add_definitions(-DVELOX_ENABLE_PARQUET)
+  if(${CMAKE_SYSTEM_NAME} MATCHES "Linux" AND (${CMAKE_SYSTEM_PROCESSOR} MATCHES
```
Maybe we should not turn it on by automatically and just leave it to user to turn it on.  We should make sure to turn it on in one of the CircleCI (probably `linux-adapters`) though

ruclz commented:
> Can we add some test? You can put the data in `tests/examples`

@Yuhta, Thanks for your comment quickly, I understand there are two methods to test IAA decompression. One is unittest, there is a ./velox/velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp file in velox. Another is velox_tpch_benchmark. Whatever which method is adopted, there is a issue need to be noted. We need to put a enough large dataset to test. Because for IAA, IAA has a job buffer initialization cost, but SW(snappy, zstd, gzip) algorithms not, so if the test dataset is too small, the bottleneck will be the IAA job buffer initialization, rather than decompression. In the local test before, I used TPC-H 100G dataset,  I can see 9%-15% performance improvement compared to ZSTD.

assignUser commented on velox/dwio/parquet/reader/CMakeLists.txt:
```diff
@@ -35,3 +35,8 @@ target_link_libraries(
   ${FMT})
+
+if(${VELOX_ENABLE_QPL})
+  target_sources(velox_dwio_native_parquet_reader PUBLIC CompressionQpl.cpp)
+  target_link_libraries(velox_dwio_native_parquet_reader qpl)
```
IIUC this PR only modifies the internals of the parquet reader in case qpl is activated right? So qpl should be linked privately as a build requirement?
```suggestion
  target_sources(velox_dwio_native_parquet_reader PRIVATE CompressionQpl.cpp)
  target_link_libraries(velox_dwio_native_parquet_reader PRIVATE qpl)
```

ruclz replied:
Thanks for your suggestion, but maybe it can`t set private in target_link_libraries statement due to conflict with over. I will modify PUBLIC to PRIVATE in target_sources statement in the later commit.

assignUser replied:
>  due to conflict with over

I am not sure I understand what you mean?

ruclz replied:
![image](https://user-images.githubusercontent.com/46258124/236385984-05250abf-007d-4b3a-b634-897b4e940e52.png)

it means that, if i set PRIVATE keyword in target_link_libraries statement, it will report this error when building.

ruclz replied:
@assignUser If you have any idea about this, it will be helpful.

assignUser replied:
Ah I see. That's not something you can fix. :+1:

ruclz commented:
> Unit test is enough to make sure it's not broken for now. You can add benchmark to track the performance later. Can you rebase on the latest version to make sure the builds are passing?

Sure!

wgtmac commented on velox/dwio/parquet/thrift/ParquetThriftTypes.h:
```diff
@@ -318,7 +318,8 @@ struct CompressionCodec {
     ZSTD = 6,
-    LZ4_RAW = 7
+    LZ4_RAW = 7,
+    QPL = 8,
```
This file is automatically generated from the `parquet.thrift` file. Adding extra codec here breaks the parquet specs. BTW, is `QPL` a codec type or implementation of one or more codec types?

ruclz replied:
@wgtmac The QPL supports de/compression compatible with the Deflate compression standard described in **RFC1951**, it can de/compress each other with zlib library if set the history buffer <= 4K and set deflate mode(not gzip or zlib mode). As for the issue you proposed, I will spend some time to find a solution to avoid adding a codec type in **parquet.thrift** file.

ruclz replied:
Parquet supports gzip, the gzip is based on deflate, the difference of gzip and deflate is window_bits which is put in **deflateInit2** function, just like decompression.

- window_bits < 0-----> deflate
- 0<window_bits<=15-----> zlib 
- window_bits>=16-----> gzip

`      constexpr int WINDOW_BITS = 15;
      // Determine if this is libz or gzip from header.
      constexpr int DETECT_CODEC = 32;
      // Initialize decompressor.
      auto ret = inflateInit2(&stream, WINDOW_BITS | DETECT_CODEC);`

ruclz replied:
Maybe not, because default gzip history buffer is 32KB, but qpl only supports history buffer <=4KB,  qpl can not decompress the file compressed by standard gzip if default history buffer. I will report to my leader and hope for some good solutions.

ruclz replied:
@wgtmac  @Yuhta  , QPL is not fully compatible with zlib now due to hardware limitation, maybe it will be solved in the future.  So looks like if we want to enable IAA for Velox, the prerequisite is we need request to add a new codec in parquet format. Am I right?  As an alternative, we can use another Intel accelerator [QAT](https://www.intel.com/content/www/us/en/architecture-and-technology/intel-quick-assist-technology-overview.html). It is fully compatible with gzip codec defined in parquet format [GZIP](https://github.com/apache/parquet-format/blob/c766945d90935ebcd4e03fee13aad2b6efcadce3/Compression.md?plain=1#L54). So it can acclerate velox parquet read without a new codec. Do you think it's a good idea to enable Velox parquet reader with QAT.

Yuhta reviewed (commented):
The test is crashing in CircleCI BTW

ruclz commented:
> The test is crashing in CircleCI BTW

Sorry, I can`t reproduce this segment fault locally, do you have any details about the CircleCI platform, including CPU, OS, GCC version and so on. Also, I run this test by executing the command velox_dwio_parquet_reader_test, not all tests, I don`t know if there is a difference.  Thanks.

Yuhta commented:
The os is Ubuntu 20.04.5 and gcc version is 9.2.1.  You can find the info in the log.  Detailed spec about CPU does not seem available though.
